### PR TITLE
Promote cost related feature gate to default true

### PR DESCRIPTION
**Pull request by cici37**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Promote cost related feature gate to default true so that the cost for cel will be calculated correctly.

The script in https://github.com/cici37/k8s-tools could be used to evaluate if your cluster would be impacted.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Promote the feature gates `StrictCostEnforcementForVAP` and `StrictCostEnforcementForWebhooks` to GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Enable feature gate `StrictCostEnforcementForVAP` and `StrictCostEnforcementForWebhooks` to true by default.
+Set the feature gates `StrictCostEnforcementForVAP` and `StrictCostEnforcementForWebhooks` to be enabled by default. These feature gates remain beta.
```

(are they beta? they're not yet [documented](https://kubernetes.io/docs/contribute/new-content/new-features/#ready-for-review-feature-gates))

**Comment by cici37** [comment]
> Changelog suggestion
> 
> ```diff
> -Enable feature gate `StrictCostEnforcementForVAP` and `StrictCostEnforcementForWebhooks` to true by default.
> +Set the feature gates `StrictCostEnforcementForVAP` and `StrictCostEnforcementForWebhooks` to be enabled by default. These feature gates remain beta.
> ```
> 
> (are they beta? they're not yet [documented](https://kubernetes.io/docs/contribute/new-content/new-features/#ready-for-review-feature-gates))

The documentation PR is https://github.com/kubernetes/website/pull/47326. Sorry for losing track on this. I'll work on merging it

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (jpbetz) [inline test/integration/apiserver/admissionwebhook/match_conditions_test.go:293]: @aaron-prindle @Jefftree Should this test be retained but with the emulation version be set to 1.31/

Comment (cici37) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:293]: note that we could not reset the feature gate to false after locking default to true. Would emulation version still work in this case?

File: pkg/features/kube_features.go
Comment (Jefftree) [inline pkg/features/kube_features.go:1012]: This should be updated to use versioned feature gates. @jpbetz can we merge  https://github.com/kubernetes/kubernetes/pull/127200 first so we won't have rebase conflicts?

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (Jefftree) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:293]: If emulation version is at 1.31 we will emulate beta behavior which still has the default to true but allows the feature to be disabled. See https://github.com/kubernetes/kubernetes/pull/127093/files#diff-8868f5a770c851b2202de4360d037910dccf41dd67cfd213baf5e58eb8e4da88R2987 to allow disablement testing after promotion to GA. We should keep the disablement tests as best as we can.

File: pkg/features/kube_features.go
Comment (cici37) [reply pkg/features/kube_features.go:1012]: Looks like 127200 also blocked by the merging of other PR currently. Is that possible to merge this PR first and I'll have a followup PR to add the deleted tests back?

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (cici37) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:293]: I guess the above setting only works for unit test since in integration test we would be using startTestServer to start the server which will set the emulation version to the default one: https://github.com/kubernetes/kubernetes/blob/3582dce11590577712f9a8db698d7e6897e23158/cmd/kube-apiserver/app/testing/testserver.go#L198

Comment (Jefftree) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:293]: Thanks for pointing this out, I'll work on a fix.

File: pkg/features/kube_features.go
Comment (Jefftree) [reply pkg/features/kube_features.go:1012]: I'll defer to @jpbetz on this, though 127200 shouldn't be blocked and just needs approval.

Comment (Jefftree) [reply pkg/features/kube_features.go:1012]: If you're looking to merge this PR first, I'd probably recommend keeping the tests and skipping them with a TODO rather than deleting and re-adding.

Comment (cici37) [reply pkg/features/kube_features.go:1012]: I added a new commit which putting back the tests with TODO comments. Will see which one merged first. :)

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (jpbetz) [inline test/integration/apiserver/admissionwebhook/match_conditions_test.go:571]: Is `t.Skip("add the tests back after the feature gate is versioned and the eumulation version in testServer is configurable")` the idiomatic way of doing this?  If https://github.com/kubernetes/kubernetes/pull/127200 doesn't merge by morning we can merge this with `t.Skip`, otherwise we can rebase and keep the tests.

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (Jefftree) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:293]: Opened https://github.com/kubernetes/kubernetes/pull/127643

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (cici37) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go]: Have switched to use test.skip for skipped tests. Thank you

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go
Comment (Jefftree) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:352]: The previous PR is merged, so you'll need to update versioned_kube_features here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/features/versioned_kube_features.go#L252-L254

and https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/features/kube_features.go#L405

Just add a new line for the GA feature gate status so we can preserve beta transition information.

Comment (cici37) [reply staging/src/k8s.io/apiserver/pkg/features/kube_features.go]: Thank you. I have updated the PR to pick up the versioned fg. Currently rebased on https://github.com/kubernetes/kubernetes/pull/127643 until it's merged.

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (Jefftree) [inline test/integration/apiserver/admissionwebhook/match_conditions_test.go:636]: @jpbetz So the issue here is that the feature gate is set before emulation version is set. I think a naive fix is just to add the emulation version check before setting the feature gate?

```
+featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))
featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.StrictCostEnforcementForWebhooks, false)
```

Comment (jpbetz) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:636]: I can live with requiring emulation version be set before feature gates during test.

If these operations are done in the wrong order it would be bad to allow the feature gates to get into a weird state, so I suspect it would be best have `SetFeatureGateEmulationVersionDuringTest` fail with a clear error.

Comment (Jefftree) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:636]: Ok @cici37 you can disregard my PR earlier and just explicitly set the emulation version before disabling the gate.

**Comment by jpbetz** [comment]
I believe the PRs this PR depends have now been merged and this PR can be rebased and updated.

**Comment by cici37** [comment]
> I believe the PRs this PR depends have now been merged and this PR can be rebased and updated.

I was watching the merge of https://github.com/kubernetes/kubernetes/pull/127643 which blocks this PR. Is there other PR I missed? BTW the tests still fail after rebased on latest commits from https://github.com/kubernetes/kubernetes/pull/127643

File: test/integration/apiserver/admissionwebhook/match_conditions_test.go
Comment (cici37) [reply test/integration/apiserver/admissionwebhook/match_conditions_test.go:638]: The test passed after set reset the version. I'll wait for https://github.com/kubernetes/kubernetes/pull/127643 to merge first.. Thank you

**Comment by Jefftree** [comment]
> > I believe the PRs this PR depends have now been merged and this PR can be rebased and updated.
> 
> I was watching the merge of #127643 which blocks this PR. Is there other PR I missed? The tests passed after the testserver fix. I guess this PR has to be merged after the merge of #127643

Can you remove the dependency on #127643 and just use this fix for the tests? https://github.com/kubernetes/kubernetes/pull/127302#discussion_r1777715525

**Comment by cici37** [comment]
@Jefftree Thanks for the work! If you are suggesting only adding
```
featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))
featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.StrictCostEnforcementForWebhooks, false)
```
without the testserver fix the tests would fail. I think we still need to pass in the emulation version into testserver. Did I miss anything?

**Comment by Jefftree** [comment]
> @Jefftree Thanks for the work! If you are suggesting only adding
> 
> ```
> featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))
> featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.StrictCostEnforcementForWebhooks, false)
> ```
> 
> without the testserver fix the tests would fail. I think we still need to pass in the emulation version into testserver. Did I miss anything?

Discussed offline. Will merge https://github.com/kubernetes/kubernetes/pull/127643, and then tests should pass.

**Comment by cici37** [comment]
Found the regression is caused by https://github.com/kubernetes/kubernetes/pull/127400.
Revert the changes in latest commit.
/cc the author @sttts :)

**Comment by liggitt** [comment]
> Found the regression is caused by #127400.

Unless this will merge ~today, please revert that separately so it can merge immediately

**Comment by liggitt** [comment]
> Found the regression is caused by #127400.

this sounds like the same risk I mentioned in https://github.com/kubernetes/kubernetes/pull/127504#discussion_r1773857134 with init-once locked CEL envs

**Comment by cici37** [comment]
The revert commit has been merged separately and this PR should be ready to merge now. Thank you!

**Comment by Jefftree** [comment]
lgtm for feature gate related changes

**Comment by cici37** [comment]
Looks like a flaky in test which is not related to current PR: https://github.com/kubernetes/kubernetes/issues/127858
/retest

**Comment by jpbetz** [comment]
@cici37 Would you update the release notes?  The code graduates to GA but the release notes say this is staying in beta.

@Jefftree would you be willing to be reviewer on this one?  Looks fine to me, but having an extra pair of eyes on feature gate changes seems appropriate.

**Comment by jpbetz** [comment]
/approve
Leaving LGTM to @Jefftree

**Comment by cici37** [comment]
> @cici37 Would you update the release notes? The code graduates to GA but the release notes say this is staying in beta.

Updated
> 
> @Jefftree would you be willing to be reviewer on this one? Looks fine to me, but having an extra pair of eyes on feature gate changes seems appropriate.

FYI Jeffrey has reviewed on the latest changes: https://github.com/kubernetes/kubernetes/pull/127302#issuecomment-2400430781

**Review by hashim21223445 — APPROVED** [review]

**Pull request merged**
